# **Airbnb Price Prediction and Neighborhood Insights using Python**

# An end-to-end data science and analytics project to explore, visualize, and model Airbnb listings with a focus on pricing trends, host behavior, and predictive modeling.


airbnb-price-prediction-insights

**IMPORTING LIBRARIES:**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import warnings
warnings.filterwarnings("ignore")



# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)  # Adjust as needed
pd.set_option('display.width', 1000)    # Stretch display width



In [3]:
data = pd.read_csv("/content/listings.csv")

In [4]:
data.shape

(2730, 79)

We got about 2730 rows with 79 columns

In [5]:
data.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name', 'description', 'neighborhood_overview', 'picture_url', 'host_id', 'host_url', 'host_name', 'host_since', 'host_location', 'host_about', 'host_response_time', 'host_response_rate', 'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood', 'host_listings_count', 'host_total_listings_count', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price', 'minimum_nights', 'maximum_nights', 'minimum_minimum_nights', 'maximum_minimum_nights', 'minimum_maximum_nights', 'maximum_maximum_nights', 'minimum_nights_avg_ntm', 'maximum_nights_avg_ntm', 'calendar_updated', 'has_availability', 'availability_30', 'availability_60', 'availability_90',
    

Here, Price is the target column. Some of the important columns to explore are neighbourhood,room_type,latitude and longitde to map,host_is_superhost,amenities,number of reviews.

In [6]:
data.head(10)

id                          listing_url       scrape_id last_scraped           source                                               name                                        description                              neighborhood_overview                                        picture_url  host_id                                   host_url     host_name  host_since        host_location                                         host_about  host_response_time host_response_rate host_acceptance_rate host_is_superhost                                 host_thumbnail_url                                   host_picture_url host_neighbourhood  host_listings_count  host_total_listings_count                host_verifications host_has_profile_pic host_identity_verified                          neighbourhood neighbourhood_cleansed  neighbourhood_group_cleansed  latitude  longitude                property_type        room_type  accommodates  bathrooms    bathrooms_text  bedrooms  beds  \
0   42515   https://www.airbnb.com/rooms/42515  20250323033821   2025-03-23      city scrape                    Geneva center - Hosting females                                                NaN                                                NaN  https://a0.muscache.com/pictures/10640277/ff1d...   185647   https://www.airbnb.com/users/show/185647        Noelle  2010-07-30  Geneva, Switzerland  Hi, I am a Management consultant professional,...  within a few hours               100%                  88%                 f  https://a0.muscache.com/im/users/185647/profil...  https://a0.muscache.com/im/users/185647/profil...                NaN                    1                          1                ['email', 'phone']                    t                      t                                    NaN      Commune de Genève                           NaN  46.20198    6.15672  Private room in rental unit     Private room             1        1.5  1.5 shared baths       1.0   1.0   
1  203997  https://www.airbnb.com/rooms/203997  20250323033821   2025-03-23      city scrape             Top location studio 1 block from lake!  Spacious studio with washer/dryer, Wi-fi, supe...  This is the most sought after area in Geneva: ...  https://a0.muscache.com/pictures/14163740/9392...  1001280  https://www.airbnb.com/users/show/1001280          Mike  2011-08-21  Geneva, Switzerland  I love airbnb, both for hosting and traveling....      within an hour               100%                 100%                 t  https://a0.muscache.com/im/pictures/user/3bbea...  https://a0.muscache.com/im/pictures/user/3bbea...                NaN                    6                         12                ['email', 'phone']                    t                      t                    Geneva, Switzerland      Commune de Genève                           NaN  46.19964    6.15580           Entire rental unit  Entire home/apt             2        1.0            1 bath       1.0   0.0   
2  276025  https://www.airbnb.com/rooms/276025  20250323033821   2025-03-23      city scrape                         Lovely flat by Geneva lake  This cozy flat is located in a very pleasant v...  Versoix is a pretty fishing village, on the ed...  https://a0.muscache.com/pictures/4156456/d3aed...  1442438  https://www.airbnb.com/users/show/1442438          Gaby  2011-11-25  Geneva, Switzerland  Hi, my name is Gaby! I'm a globetrotter. I spe...      within an hour               100%                 100%                 t  https://a0.muscache.com/im/users/1442438/profi...  https://a0.muscache.com/im/users/1442438/profi...                NaN                    2                          2                ['email', 'phone']                    t                      t           Versoix, Geneva, Switzerland                Versoix                           NaN  46.28031    6.16833           Entire rental unit  Entire home/apt             3        1.0            1 bath       1.0   4.0   
3  338682  https://www.airbnb.com/r

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2730 entries, 0 to 2729
Data columns (total 79 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            2730 non-null   int64  
 1   listing_url                                   2730 non-null   object 
 2   scrape_id                                     2730 non-null   int64  
 3   last_scraped                                  2730 non-null   object 
 4   source                                        2730 non-null   object 
 5   name                                          2730 non-null   object 
 6   description                                   2636 non-null   object 
 7   neighborhood_overview                         1022 non-null   object 
 8   picture_url                                   2730 non-null   object 
 9   host_id                                       2730 non-null   i

Data contains many null values,contains both numeric and object data type.Our target columns has some null values as wll the type is object.

In [8]:
data.describe()

id     scrape_id       host_id  host_listings_count  host_total_listings_count  neighbourhood_group_cleansed     latitude    longitude  accommodates    bathrooms     bedrooms         beds  minimum_nights  maximum_nights  minimum_minimum_nights  maximum_minimum_nights  minimum_maximum_nights  maximum_maximum_nights  minimum_nights_avg_ntm  maximum_nights_avg_ntm  calendar_updated  availability_30  availability_60  availability_90  availability_365  number_of_reviews  number_of_reviews_ltm  number_of_reviews_l30d  availability_eoy  number_of_reviews_ly  estimated_occupancy_l365d  estimated_revenue_l365d  review_scores_rating  review_scores_accuracy  review_scores_cleanliness  review_scores_checkin  review_scores_communication  review_scores_location  review_scores_value  calculated_host_listings_count  calculated_host_listings_count_entire_homes  calculated_host_listings_count_private_rooms  calculated_host_listings_count_shared_rooms  reviews_per_month
count  2.730000e+03  2.730000e+03  2.730000e+03          2730.000000                2730.000000                           0.0  2730.000000  2730.000000   2730.000000  2011.000000  2519.000000  2007.000000     2730.000000     2730.000000             2730.000000             2730.000000             2730.000000             2730.000000             2730.000000             2730.000000               0.0      2730.000000      2730.000000      2730.000000       2730.000000        2730.000000            2730.000000             2730.000000       2730.000000           2730.000000                2730.000000              2005.000000           2057.000000             2057.000000                2057.000000            2057.000000                  2057.000000             2057.000000          2057.000000                     2730.000000                                  2730.000000                                   2730.000000                                  2730.000000        2057.000000
mean   5.964151e+17  2.025032e+13  1.848724e+08            34.445788                  46.741758                           NaN    46.207117     6.144779      2.641758     1.178767     1.255260     1.562033        9.124176      426.071795                8.441758                9.348718              547.061538              575.331868                8.997253              566.341392               NaN         9.478022        21.692674        34.686081        134.232234          26.465568               6.363370                0.355678        108.926740              6.250183                  40.411722              6160.074813              4.732188                4.765843                   4.718935               4.811006                     4.802504                4.792825             4.619208                       18.287912                                    16.863370                                      1.351282                                     0.003663           0.970632
std    5.442481e+17  5.547891e-01  1.972699e+08            94.648896                 120.071377                           NaN     0.019881     0.025079      1.525361     0.492977     0.999775     1.110995       46.023211      403.141721               44.882499               45.292656              451.374369              447.990797               45.044895              444.715760               NaN        10.381883        22.210650        34.365622        134.615069          58.965950              14.020604                1.126911        106.102655             14.250290                  67.312290             11085.035780              0.396831                0.394143                   0.407506               0.368086                     0.379760                0.337144             0.436437                       45.814872                                    45.499041                                      3.392927                                     0.071531           1.389309
min    4.251500e+04  2.025032e+13  6.783700e+04             1.000000                 

In [9]:
data["price"]

0        $89.00
1       $128.00
2        $70.00
3           NaN
4        $55.00
         ...   
2725     $85.00
2726     $80.00
2727    $354.00
2728    $262.00
2729    $252.00
Name: price, Length: 2730, dtype: object

In [10]:
data["price"] = data["price"].replace(r'[^\d.]', '', regex=True).astype(float)


In [11]:
data["price"]

0        89.0
1       128.0
2        70.0
3         NaN
4        55.0
        ...  
2725     85.0
2726     80.0
2727    354.0
2728    262.0
2729    252.0
Name: price, Length: 2730, dtype: float64

In [12]:
data["price"].describe()

count    2005.000000
mean      153.721696
std       239.834562
min        21.000000
25%        82.000000
50%       115.000000
75%       166.000000
max      8214.000000
Name: price, dtype: float64

The average price is 153.72 CHF.Valis listings are 2005.STD shows widely spread. min denotes cheap listings i.e private or shared rooms. max likely be the outlier costing around 8214. The mean is higher than min  i.e right skewed.

In [13]:
data.isnull().sum().sort_values(ascending=False)


neighbourhood_group_cleansed                    2730
calendar_updated                                2730
license                                         2729
host_neighbourhood                              2700
neighborhood_overview                           1708
neighbourhood                                   1708
host_about                                      1405
estimated_revenue_l365d                          725
price                                            725
beds                                             723
bathrooms                                        719
host_response_rate                               682
host_response_time                               682
review_scores_cleanliness                        673
review_scores_checkin                            673
review_scores_location                           673
first_review                                     673
review_scores_communication                      673
review_scores_rating                             673
review_scores_accuracy                           673
review_scores_value                              673
reviews_per_month                                673
last_review                                      673
host_location                                    517
host_acceptance_rate                             461
bedrooms                                         211
description                                       94
has_availability                                  68
host_is_superhost                                 27
bathrooms_text                                     3
scrape_id                                          0
host_name                                          0
host_since                                         0
host_id                                            0
host_url                                           0
picture_url                                        0
last_scraped                                       0
source                                             0
name                                               0
id                                                 0
listing_url                                        0
latitude                                           0
neighbourhood_cleansed                             0
host_listings_count                                0
host_picture_url                                   0
host_total_listings_count                          0
host_thumbnail_url                                 0
maximum_maximum_nights                             0
minimum_maximum_nights                             0
maximum_minimum_nights                             0
minimum_minimum_nights                             0
maximum_nights                                     0
minimum_nights                                     0
amenities                                          0
longitude                                          0
room_type                                          0
property_type                                      0
accommodates                                       0
host_verifications                                 0
host_identity_verified                             0
host_has_profile_pic                               0
availability_30                                    0
maximum_nights_avg_ntm                             0
estimated_occupancy_l365d                          0
number_of_reviews_ly                               0
availability_eoy                                   0
number_of_reviews_l30d                             0
number_of_reviews                                  0
number_of_reviews_ltm                              0
calendar_last_scraped                              0
availability_365                                   0
minimum_nights_avg_ntm                             0
availability_90                                    0
availability_60                                    0
instant_bookable                                   0
calculated_host_listings_

We are dropping columns like host_neighbourhood almost all missing and not critical.host_about personal information are not useful.neighborhood_overview / neighbourhood neighbourhood_cleansed instead,
estimated_revenue_l365d,not needed for our model.
first_review, last_review has dates not needed for price modeling.scrape_id, host_name	IDs are irrelevant for analysis.


In [14]:
data['neighbourhood_cleansed'].isnull().sum()


np.int64(0)

In [16]:
for col in data.columns:
    print(f"\n🔹 {col} ({data[col].nunique()} unique values):")
    print(data[col].unique()[:10])  # shows first 10 unique values



🔹 id (2730 unique values):
[ 42515 203997 276025 338682 399388 419631 626512 649023 664917 693976]

🔹 listing_url (2730 unique values):
['https://www.airbnb.com/rooms/42515'
 'https://www.airbnb.com/rooms/203997'
 'https://www.airbnb.com/rooms/276025'
 'https://www.airbnb.com/rooms/338682'
 'https://www.airbnb.com/rooms/399388'
 'https://www.airbnb.com/rooms/419631'
 'https://www.airbnb.com/rooms/626512'
 'https://www.airbnb.com/rooms/649023'
 'https://www.airbnb.com/rooms/664917'
 'https://www.airbnb.com/rooms/693976']

🔹 scrape_id (1 unique values):
[20250323033821]

🔹 last_scraped (1 unique values):
['2025-03-23']

🔹 source (2 unique values):
['city scrape' 'previous scrape']

🔹 name (2667 unique values):
['Geneva center - Hosting females'
 'Top location studio 1 block from lake!' 'Lovely flat by Geneva lake'
 'Top location 1 block from the lake!' 'Nice modern bedroom with balcony'
 "Beautiful Apm't in central Geneva (NB Guest Room)"
 'Design apartment 2 blocks from lake'
 'Artisti

In [17]:
data.drop(columns=[
    'host_neighbourhood', 'host_about', 'neighborhood_overview', 'neighbourhood',
    'estimated_revenue_l365d', 'first_review', 'last_review',
    'scrape_id', 'host_name'
], inplace=True)


Dropping all the non-crictical columns

In [18]:
data.dropna(subset=['price', 'beds', 'bathrooms', 'bedrooms', 'host_is_superhost'], inplace=True)


Dropping the NULL values in the most crictical columns.

In [20]:
data.shape

(1969, 70)

In [21]:
data.isnull().sum().sort_values(ascending=False)

calendar_updated                                1969
neighbourhood_group_cleansed                    1969
license                                         1968
review_scores_accuracy                           482
review_scores_location                           482
review_scores_value                              482
review_scores_communication                      482
review_scores_cleanliness                        482
reviews_per_month                                482
review_scores_rating                             482
review_scores_checkin                            482
host_location                                    375
host_response_rate                               219
host_response_time                               219
host_acceptance_rate                             132
description                                       52
has_availability                                  22
name                                               0
id                                                 0
source                                             0
last_scraped                                       0
listing_url                                        0
host_identity_verified                             0
neighbourhood_cleansed                             0
latitude                                           0
longitude                                          0
property_type                                      0
host_since                                         0
host_id                                            0
host_url                                           0
host_is_superhost                                  0
host_thumbnail_url                                 0
host_picture_url                                   0
host_listings_count                                0
host_total_listings_count                          0
host_verifications                                 0
host_has_profile_pic                               0
picture_url                                        0
minimum_minimum_nights                             0
maximum_nights                                     0
minimum_nights                                     0
price                                              0
amenities                                          0
beds                                               0
bedrooms                                           0
bathrooms_text                                     0
bathrooms                                          0
accommodates                                       0
room_type                                          0
maximum_minimum_nights                             0
minimum_nights_avg_ntm                             0
maximum_maximum_nights                             0
maximum_nights_avg_ntm                             0
minimum_maximum_nights                             0
availability_eoy                                   0
number_of_reviews_l30d                             0
number_of_reviews_ltm                              0
number_of_reviews                                  0
calendar_last_scraped                              0
availability_365                                   0
availability_90                                    0
availability_60                                    0
availability_30                                    0
number_of_reviews_ly                               0
estimated_occupancy_l365d                          0
instant_bookable                                   0
calculated_host_listings_count                     0
calculated_host_listings_count_entire_homes        0
calculated_host_listings_count_private_rooms       0
calculated_host_listings_count_shared_rooms        0
dtype: int64

In [22]:
data.drop(columns=['calendar_updated', 'neighbourhood_group_cleansed', 'license'], inplace=True)


Dropping the columns containing almost NULL values

**IMPUTATION WITH MEDIAN:**

In [25]:
review_cols=[ 'review_scores_accuracy', 'review_scores_location', 'review_scores_value',
    'review_scores_communication', 'review_scores_cleanliness',
    'review_scores_checkin', 'review_scores_rating']

In [26]:
data[review_cols] = data[review_cols].fillna(data[review_cols].median())

For the column review_per_month Nan indicates listings has no bookings. Therefore filling NaN with 0 is accurate and honest.

In [27]:
data['reviews_per_month'].fillna(0, inplace=True)


For the columns host response rate and host acceptance rate the values are in % in object type.So converting those columns to float and filling the null values with median.

In [28]:
# Strip % and convert to float
data['host_response_rate'] = data['host_response_rate'].str.rstrip('%').astype(float)
data['host_acceptance_rate'] = data['host_acceptance_rate'].str.rstrip('%').astype(float)

# Fill remaining nulls with median
data['host_response_rate'].fillna(data['host_response_rate'].median(), inplace=True)
data['host_acceptance_rate'].fillna(data['host_acceptance_rate'].median(), inplace=True)


For the text columns like host response time,host location and description, the Null values are filled with unknown and no description.

In [29]:
data['host_location'].fillna('Unknown', inplace=True)
data['host_response_time'].fillna('Unknown', inplace=True)
data['description'].fillna('No description', inplace=True)


In [30]:
data['has_availability'] = data['has_availability'].map({'t': True, 'f': False})
data['has_availability'].fillna(False, inplace=True)


In [31]:
data.isnull().sum().sort_values(ascending=False)

id                                              0
listing_url                                     0
last_scraped                                    0
source                                          0
name                                            0
description                                     0
picture_url                                     0
host_id                                         0
host_url                                        0
host_since                                      0
host_location                                   0
host_response_time                              0
host_response_rate                              0
host_acceptance_rate                            0
host_is_superhost                               0
host_thumbnail_url                              0
host_picture_url                                0
host_listings_count                             0
host_total_listings_count                       0
host_verifications                              0
host_has_profile_pic                            0
host_identity_verified                          0
neighbourhood_cleansed                          0
latitude                                        0
longitude                                       0
property_type                                   0
room_type                                       0
accommodates                                    0
bathrooms                                       0
bathrooms_text                                  0
bedrooms                                        0
beds                                            0
amenities                                       0
price                                           0
minimum_nights                                  0
maximum_nights                                  0
minimum_minimum_nights                          0
maximum_minimum_nights                          0
minimum_maximum_nights                          0
maximum_maximum_nights                          0
minimum_nights_avg_ntm                          0
maximum_nights_avg_ntm                          0
has_availability                                0
availability_30                                 0
availability_60                                 0
availability_90                                 0
availability_365                                0
calendar_last_scraped                           0
number_of_reviews                               0
number_of_reviews_ltm                           0
number_of_reviews_l30d                          0
availability_eoy                                0
number_of_reviews_ly                            0
estimated_occupancy_l365d                       0
review_scores_rating                            0
review_scores_accuracy                          0
review_scores_cleanliness                       0
review_scores_checkin                           0
review_scores_communication                     0
review_scores_location                          0
review_scores_value                             0
instant_bookable                                0
calculated_host_listings_count                  0
calculated_host_listings_count_entire_homes     0
calculated_host_listings_count_private_rooms    0
calculated_host_listings_count_shared_rooms     0
reviews_per_month                               0
dtype: int64

In [32]:
data.shape

(1969, 67)

In [34]:
data.to_csv("geneva_airbnb_cleaned.csv", index=False)
